crates/rspack_napi/src/by_ref.rs

```diff
@@ -0,0 +1,107 @@
+use std::{mem::ManuallyDrop, ops::Deref};
+
+use derivative::Derivative;
+use napi::{
+  bindgen_prelude::{FromNapiValue, ToNapiValue, TypeName, ValidateNapiValue},
+  Env, JsFunction, JsNumber, Ref,
+};
+
+/// Wrapper for napi types that provides by-reference equality and hash
+#[derive(Derivative, Debug, Clone)]
+#[derivative(Eq, PartialEq, Hash)]
+pub struct ByRef<T> {
```

Comment:
I didn't see why this is necessary. Would you please make some clarifications?

Reply:
This makes [the tsfn in `JsFilename`]( https://github.com/branchseer/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_binding_values/src/filename.rs#L20 ) implements `Hash` and `Eq`, allowing it to be [converted to `Arc<dyn FilenameFn>`](https://github.com/branchseer/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_binding_values/src/filename.rs#L41).  `FilenameFn` requires `Hash` and `Eq` because that `Filename` needs to be hashmap keys [here](https://github.com/web-infra-dev/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_plugin_runtime/src/runtime_module/get_chunk_filename.rs#L122) (this is consistent with the [webpack implementation](https://github.com/webpack/webpack/blob/0bc85d127c14347d646f1354a6cbf2d1142905d8/lib/runtime/GetChunkFilenameRuntimeModule.js#L48))

Reply:
It still seems not necessary to implement this for filename, a filename enum can accept both string and function variant looks enough. Webpack didn't seem necessary doing anything with its being the `HashMap` key, you can implement this with other struct roughly like a `Vec<(Filename, Chunks)>`.

Reply:
Webpack is grouping chunks by filenames [here](https://github.com/webpack/webpack/blob/0bc85d127c14347d646f1354a6cbf2d1142905d8/lib/runtime/GetChunkFilenameRuntimeModule.js#L60-L64). And currently rspack [does the same](https://github.com/web-infra-dev/rspack/blob/4ccff45dde2ff9b26649ce99ef95ec41525428ad/crates/rspack_plugin_runtime/src/runtime_module/get_chunk_filename.rs#L135-L139), and later in [here](https://github.com/web-infra-dev/rspack/blob/4ccff45dde2ff9b26649ce99ef95ec41525428ad/crates/rspack_plugin_runtime/src/runtime_module/get_chunk_filename.rs#L242-L249) the grouped chunks are iterated to generated `static_urls`, which is also grouped by `filename`.

I don't see how grouping is possible if `Filename` doesn't impl `Eq`, unless we only group by filename strings and consider all filename functions distinct to each other?